From 2fc292e80fcb0e81cbb379fbe242c6797b0903eb Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <uwu@icenowy.me>
Date: Sun, 19 Jan 2025 09:00:21 +0800
Subject: [PATCH 7/8] zink: skip empty VkSubmitInfo when submitting

The closed blob from Imagination for their Rogue GPUs cannot accept
empty VkSubmitInfo, skip them.

Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
---
 src/gallium/drivers/zink/zink_batch.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/gallium/drivers/zink/zink_batch.c b/src/gallium/drivers/zink/zink_batch.c
index b8abb68e53d..137d94eb6fa 100644
--- a/src/gallium/drivers/zink/zink_batch.c
+++ b/src/gallium/drivers/zink/zink_batch.c
@@ -671,6 +671,13 @@ submit_queue(void *data, void *gdata, int thread_index)
       }
    }
 
+   if (si[ZINK_SUBMIT_WAIT_ACQUIRE].waitSemaphoreCount != 0 &&
+       si[ZINK_SUBMIT_WAIT_FD].waitSemaphoreCount == 0) {
+      si[ZINK_SUBMIT_WAIT_FD] = si[ZINK_SUBMIT_WAIT_ACQUIRE];
+      num_si--;
+      submit++;
+   }
+
    /* then the real submit */
    si[ZINK_SUBMIT_CMDBUF].waitSemaphoreCount = util_dynarray_num_elements(&bs->wait_semaphores, VkSemaphore);
    si[ZINK_SUBMIT_CMDBUF].pWaitSemaphores = bs->wait_semaphores.data;
-- 
2.49.0

